File: 백준/Gold/5557. 1학년/1학년.js
const input = require("fs").readFileSync("/dev/stdin", "utf-8").trim().split("\n");

const n = Number(input[0]);
const nums = input[1].trim().split(" ").map(Number);

const dp = Array.from({ length: n - 1 }, () => Array(21).fill(BigInt(0)));
dp[0][nums[0]] = BigInt(1);

for (let i = 1; i < n - 1; i++) {
	for (let val = 0; val < 21; val++) {
		if (dp[i - 1][val] > 0) {
			const plus = val + nums[i];
			const minus = val - nums[i];

			if (0 <= plus && plus <= 20) dp[i][plus] += dp[i - 1][val];
			if (0 <= minus && minus <= 20) dp[i][minus] += dp[i - 1][val];
		}
	}
}

console.log(String(dp[n - 2][nums[n - 1]]));

File: 백준/Gold/5557. 1학년/1학년.py

# dp - 메모이제이션

n = int(input())
nums = list(map(int,input().split()))

dp = [[0] * 21 for _ in range(n - 1)] # 마지막 숫자는 결과여야되니까
dp[0][nums[0]] = 1

for i in range(1, n - 1):
    for val in range(21):
        if dp[i - 1][val]:
            plus = val + nums[i]
            minus = val - nums[i]
            if 0 <= plus <= 20:
                dp[i][plus] += dp[i - 1][val]
            if 0 <= minus <= 20:
                dp[i][minus] += dp[i -1][val]

print(dp[n - 2][nums[-1]])